File: arti-coding-dojo/000-roman-numerals/main.go
package main

import (
	"fmt"
	"strings"
)

type convert func(int) int

func calc(symbol string, symbolNumber int, inputNumber int) int {
	count := inputNumber / symbolNumber
	fmt.Print(strings.Repeat(symbol, count))
	return inputNumber % symbolNumber
}

func roman(symbol string, symbolNumber int) convert {
	return func(inputNumber int) int {
		return calc(symbol, symbolNumber, inputNumber)
	}
}

var setting = []convert{
	roman("M", 1000),
	roman("CM", 900),
	roman("D", 500),
	roman("CD", 400),
	roman("C", 100),
	roman("XC", 90),
	roman("L", 50),
	roman("XL", 40),
	roman("X", 10),
	roman("IX", 9),
	roman("V", 5),
	roman("IV", 4),
	roman("I", 1),
	func(inputNumber int) int {
		fmt.Println("")
		return inputNumber
	},
}

func printRoman(inputNumber int) {
	res := inputNumber
	for _, belt := range setting {
		res = belt(res)
	}
}

File: arti-coding-dojo/000-roman-numerals/main.js
const print = (str) => process.stdout.write(str);
const chain = (a, b) => b(a);

function calc(sign, signNumber, inputNumber) {
    const count = Math.floor(inputNumber / signNumber);
    print(sign.repeat(count));
    return inputNumber % signNumber;
}

const roman = (sign, signNumber) => calc.bind(null, sign, signNumber);

const define = [
    (num) => { print(`${num} : `); return num; },
    roman("M", 1000),
    roman("CM", 900),
    roman("D", 500),
    roman("CD", 400),
    roman("C", 100),
    roman("XC", 90),
    roman("L", 50),
    roman("XL", 40),
    roman("X", 10),
    roman("V", 5),
    roman("IV", 4),
    roman("I", 1),
    () => console.log(),
];

const printRoman = (num) => define.reduce(chain, num);

printRoman(1954);
printRoman(1776);
printRoman(1066);
printRoman(207);
printRoman(246);
printRoman(39);
printRoman(10);
printRoman(7);
printRoman(4);
printRoman(3);
printRoman(2);
printRoman(1);

